design/004-IoT-Adoption-Watcher.md

```diff
@@ -0,0 +1,248 @@
+was described in [perun-proposal#003](./003-IoT-Adoption.md). Detailed
+description of the functionality and design for implementing the watching
+component is presented in this proposal.
+
+
+## Motivation
+
+<!-- introduction to the problem being solved & its background.  -->
+
+In the current implementation of the framework (go-perun v0.6.0), the watcher is
+integrated into `client.Channel` type and an instance must be started for each
+channel.
+
+However, it is desirable to run the watcher as a separate component for IoT use
+cases. Because, for the perun protocol to work, the watcher should be actively
+watching the blockchain for any states being registered and; if the registered
+state was not the latet off-chain state, the watcher must immediately register
+the latest off-chain state.  Given the connectivity and power constraints of
+IoT devices, it might not be possible to meet these requirements if watching
+service is running on the IoT device.
+
+The current design cannot be extended to implement a watcher that can run
+as a remote service. Hence, a new design is proposed where, the watcher can be run:
+1. As a part of go-perun client itself (ideal for capable hardware).
+2. As an independant component on the same or different computer (ideal for
+   constrained hardware).
+
+## Details
+
+<!-- Provide a detailed description of the proposal. -->
+
+### Functionalities of the watcher
+
+1. User should be able to initialize the watcher without associating it to any
+   particular channel. After initialization, it must have access to a wallet
+   containing keys for an on-chain account with sufficient funds (to pay
+   transaction fees when registering off-chain states).
+
+2. Main component should be able to register/deregister channel IDs to
+   start/stop watching for on-chain events.
+
+3. After a channel is registered,
+
+   1. Main component should be able to periodically send the newer off-chain
+      states.
+
+   2. Main component should be notified by the watcher, when a state is
+      registered or progressed on the blockchain.
+
+4. If the state registered on the blockchain and is not the latest off-chain
+   state known to the watcher, then it must register the latest state on the
+   blockchain.
+
+### Interfaces of the watcher
+
+In order to realize the desired functionalities, watcher should provide the
+following interfaces:
+
+
+```go
+type Watcher struct{}
+
+// For initalizing the watcher.
+func NewWatcher(rs RegisterSubscriber) *Watcher {...}
+
+// For interacting with the blockchain.
+type RegisterSubscriber interface {
+    Subscriber
+    Registerer
+}
+
+// For registering off-chain state for a given ledger channel,
+// along with the off-chain states of its sub-channels and virtual channels.
+type Registerer interface {
+    Register(State, SubStates)
+}
+
+// For receiving registered and progressed events for a given channel ID.
+type Subscribe interface {
```

Comment:
Is there a typo? Should this be Subscribe**r** instead of Subscribe?

Reply:
Yes, it should be `Subscriber`. Have fixed it in efb39d7.